lexicon/providers/hetzner.py

```diff
@@ -0,0 +1,604 @@
+        if len(identifier) > 7:
+            parts = identifier.split('/')
+            rdtype, name, content = parts[0], parts[1], '/'.join(parts[2:])
+        else:
+            records = self.list_records()
+            for record in records:
+                if record['id'] == identifier:
+                    rdtype, name, content = record['type'], record['name']+'.', record['content']
+        return rdtype, name, content
+
+    def _wellformed_content(self, rdtype, content):
+        if rdtype == 'TXT':
+            if content[0] != '"':
+                content = '"' + content
+            if content[-1] != '"':
+                content += '"'
+        if rdtype in ('CNAME', 'MX', 'NS', 'SRV'):
+            if content[-1] != '.':
+                content = self._fqdn_name(content)
+        return content
+
+    def _raw_content(self, rdtype, content):
+        if rdtype == 'TXT':
+            content = content.strip('"')
+        return content
+
+    def _concatenate(self):
+        action = self._get_lexicon_option('action')
+        identifier = self._get_lexicon_option('identifier')
+        rdtype = self._get_lexicon_option('type')
+        name = self._get_lexicon_option('name')
+        concatenate = True if self._get_provider_option('concatenate') == 'yes' else False
+        name_update = name
+        if identifier:
+            rdtype, name, _ = self._parse_identifier(identifier)
+            name_update = self._fqdn_name(name_update) if name_update else name
+        if action != 'list' and rdtype and rdtype != 'CNAME' and name and concatenate:
+            if action != 'update' or name == name_update:
+                LOGGER.info('Hetzner => Enabled CNAME lookup, '
+                            'see --concatenate option with \'lexicon hetzner --help\'')
+                return name, True
+        LOGGER.info('Hetzner => Disabled CNAME lookup, '
+                    'see --concatenate option with \'lexicon hetzner --help\'')
+        return name, False
+
+    def _propagated(self, rdtype, name, content):
+        propagated = True if self._get_provider_option('propagated') == 'yes' else False
+        if propagated:
+            retry, max_retry = 0, 20
+            while retry < max_retry:
+                for rdata in self._dns_lookup((self.cname if self.cname
+                                               else self._fqdn_name(name)),
+                                              rdtype, self.nameservers):
+                    if self._wellformed_content(rdtype, content) == rdata.to_text():
+                        return True
+                retry += 1
+                LOGGER.info('Hetzner => Record is not propagated, %d retries remaining - '
+                            'wait 30s...', (max_retry - retry))
+                time.sleep(30)
+        return False
+
+    # DNS Helpers
+    ###########################################################################
+    def _dns_lookup(self, qname, rdtype, nameservers=None):
```

Comment:
For this matter, in the auto provider, I made a direct call to the nslookup binary. I do not know if it is better to add an optional dependency to dnspython and call it, or to use nslookup.

But we should choose an approach, and be in sync throughout lexicon code.

Reply:
I need dnspython for zone manipulating anyway. And with dnspython I know how to use another nameserver as resolver for dns lookups, what - I think - gives me the opportunity to get the propagated state for new records faster.

Reply:
If you need it to manipulate the zone, then it will be the way. However, relying on another dns server than the one set for the host could be a problem. In a corporate protected network, you are not sure that you could touch an external dns server appart the internal one set by the network administrators.

Reply:
Maybe I should use the local resolver as the initial one...